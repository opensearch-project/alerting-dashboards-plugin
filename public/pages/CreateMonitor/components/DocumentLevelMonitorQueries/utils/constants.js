/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { OPERATORS_MAP } from '../../MonitorExpressions/expressions/utils/constants';

export const DOC_LEVEL_INPUT_FIELD = 'doc_level_input';

/**
 * A list of the operators currently supported for defining queries through the UI.
 */
export const SUPPORTED_DOC_LEVEL_QUERY_OPERATORS = [
  OPERATORS_MAP.IS.value,
  OPERATORS_MAP.IS_NOT.value,
  OPERATORS_MAP.IS_GREATER.value,
  OPERATORS_MAP.IS_GREATER_EQUAL.value,
  OPERATORS_MAP.IS_LESS.value,
  OPERATORS_MAP.IS_LESS_EQUAL.value,
];

/**
 * Initial implementation of document level monitors used these feature-specific operator values.
 * Refactored document level monitors assets to reuse the OPERATORS_MAP to reduce duplicate code
 * as new query types become supported.
 */
export const LEGACY_QUERY_OPERATORS = {
  IS: { text: 'is', value: '==' },
  IS_NOT: { text: 'is not', value: '!=' },
};

/**
 * These patterns delineate the field being queries from the query value
 * when using query string query syntax.
 */
export const QUERY_STRING_QUERY_OPERATORS = {
  [OPERATORS_MAP.IS_GREATER.value]: ':>',
  [OPERATORS_MAP.IS_GREATER_EQUAL.value]: ':>=',
  [OPERATORS_MAP.IS_LESS.value]: ':<',
  [OPERATORS_MAP.IS_LESS_EQUAL.value]: ':<=',
};

/**
 * Similar to OPERATORS_QUERY_MAP, this const contains the query patterns
 * for the SUPPORTED_DOC_LEVEL_QUERY_OPERATORS.
 */
export const DOC_LEVEL_QUERY_MAP = {
  [LEGACY_QUERY_OPERATORS.IS.value]: {
    query: ({ field, query }) => `${field}:\"${query}\"`,
  },
  [LEGACY_QUERY_OPERATORS.IS_NOT.value]: {
    query: ({ field, query }) => `NOT (${field}:\"${query}\")`,
  },
  [OPERATORS_MAP.IS.value]: {
    query: ({ field, query }) => `${field}:\"${query}\"`,
  },
  [OPERATORS_MAP.IS_NOT.value]: {
    query: ({ field, query }) => `NOT (${field}:\"${query}\")`,
  },
  [OPERATORS_MAP.IS_GREATER.value]: {
    query: ({ field, query }) =>
      `${field}${QUERY_STRING_QUERY_OPERATORS[OPERATORS_MAP.IS_GREATER.value]}${query}`,
  },
  [OPERATORS_MAP.IS_GREATER_EQUAL.value]: {
    query: ({ field, query }) =>
      `${field}${QUERY_STRING_QUERY_OPERATORS[OPERATORS_MAP.IS_GREATER_EQUAL.value]}${query}`,
  },
  [OPERATORS_MAP.IS_LESS.value]: {
    query: ({ field, query }) =>
      `${field}${QUERY_STRING_QUERY_OPERATORS[OPERATORS_MAP.IS_LESS.value]}${query}`,
  },
  [OPERATORS_MAP.IS_LESS_EQUAL.value]: {
    query: ({ field, query }) =>
      `${field}${QUERY_STRING_QUERY_OPERATORS[OPERATORS_MAP.IS_LESS_EQUAL.value]}${query}`,
  },
};
