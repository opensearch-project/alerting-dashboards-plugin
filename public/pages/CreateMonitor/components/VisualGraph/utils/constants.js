/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

export const SIZE_RANGE = [3, 3];
export const LINE_STYLES = { strokeWidth: 1 };
export const ANNOTATION_STYLES = { strokeWidth: 1, stroke: 'red' };
export const HINT_STYLES = {
  background: '#3a3a48',
  borderRadius: '5px',
  padding: '7px 10px',
  fontSize: '12px',
  fontWeight: 'bold',
  color: '#fff',
  textAlign: 'left',
  whiteSpace: 'nowrap',
  boxShadow: '0 2px 4px rgba(0,0,0,.5)',
};

// Don't want values to reach top of graph so increase by a buffer % so we always have a yDomain with a slightly higher max
export const Y_DOMAIN_BUFFER = 1.4; // 40%

export const X_DOMAIN_BUFFER = 0.2; // 40%

// Size of circle for each point on graph
export const DEFAULT_MARK_SIZE = 3;

export const BAR_PERCENTAGE = 0.015;

export const BAR_KEY_COUNT = 3;
