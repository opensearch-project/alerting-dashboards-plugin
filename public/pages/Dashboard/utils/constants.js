/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const EMPTY_ALERT_LIST = {
  ACTIVE: 0,
  ACKNOWLEDGED: 0,
  ERROR: 0,
  total: 0,
  alerts: [],
};

export const MAX_ALERT_COUNT = 10000;

export const DEFAULT_PAGE_SIZE_OPTIONS = [5, 10, 20, 50];
export const DEFAULT_GET_ALERTS_QUERY_PARAMS = {
  alertState: 'ALL',
  from: 0,
  search: '',
  severityLevel: 'ALL',
  size: 20,
  sortDirection: 'desc',
  sortField: 'start_time',
};

export const DEFAULT_GET_ALERTS_QUERY_PARAMS_BY_TRIGGER = {
  alertState: 'ALL',
  from: 0,
  search: '',
  severityLevel: 'ALL',
  size: MAX_ALERT_COUNT,
  sortDirection: 'desc',
  sortField: 'start_time',
};
