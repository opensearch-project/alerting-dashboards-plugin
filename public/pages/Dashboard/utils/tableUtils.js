/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import _ from 'lodash';
import { EuiLink } from '@elastic/eui';
import moment from 'moment';
import { ALERT_STATE, DEFAULT_EMPTY_DATA } from '../../../utils/constants';
import { PLUGIN_NAME } from '../../../../utils/constants';

const renderTime = (time) => {
  const momentTime = moment(time);
  if (time && momentTime.isValid()) return momentTime.format('MM/DD/YY h:mm a');
  return DEFAULT_EMPTY_DATA;
};

export const queryColumns = [
  {
    field: 'start_time',
    name: 'Alert start time',
    sortable: true,
    truncateText: false,
    render: renderTime,
    dataType: 'date',
  },
  {
    field: 'end_time',
    name: 'Alert end time',
    sortable: true,
    truncateText: false,
    render: renderTime,
    dataType: 'date',
  },
  {
    field: 'trigger_name',
    name: 'Trigger name',
    sortable: true,
    truncateText: true,
    textOnly: true,
  },
  {
    field: 'severity',
    name: 'Severity',
    sortable: false,
    truncateText: false,
  },
  {
    field: 'state',
    name: 'State',
    sortable: false,
    truncateText: false,
    render: (state, alert) => {
      const stateText =
        typeof state !== 'string' ? DEFAULT_EMPTY_DATA : _.capitalize(state.toLowerCase());
      return state === ALERT_STATE.ERROR ? `${stateText}: ${alert.error_message}` : stateText;
    },
  },
  {
    field: 'acknowledged_time',
    name: 'Time acknowledged',
    sortable: true,
    truncateText: false,
    render: renderTime,
    dataType: 'date',
  },
];

export const bucketColumns = [
  {
    field: 'start_time',
    name: 'Alert start time',
    sortable: true,
    truncateText: false,
    render: renderTime,
    dataType: 'date',
  },
  {
    field: 'end_time',
    name: 'Alert last updated',
    sortable: true,
    truncateText: false,
    render: (endTime, alert) => {
      const ackTime = alert.acknowledged_time;
      return renderTime(Math.max(endTime, ackTime));
    },
    dataType: 'date',
  },
  {
    field: 'state',
    name: 'State',
    sortable: false,
    truncateText: false,
    render: (state, alert) => {
      const stateText =
        typeof state !== 'string' ? DEFAULT_EMPTY_DATA : _.capitalize(state.toLowerCase());
      return state === ALERT_STATE.ERROR ? `${stateText}: ${alert.error_message}` : stateText;
    },
  },
  {
    field: 'trigger_name',
    name: 'Trigger name',
    sortable: true,
    truncateText: true,
    textOnly: true,
  },
  {
    field: 'severity',
    name: 'Severity',
    sortable: false,
    truncateText: false,
  },
];

export const alertColumns = (
  history,
  httpClient,
  loadingMonitors,
  location,
  monitors,
  notifications,
  setFlyout,
  openFlyout,
  closeFlyout,
  refreshDashboard
) => [
  {
    field: 'total',
    name: 'Alerts',
    sortable: true,
    truncateText: false,
    render: (total, alert) => {
      return (
        <EuiLink
          onClick={() => {
            openFlyout({
              ...alert,
              history,
              httpClient,
              loadingMonitors,
              location,
              monitors,
              notifications,
              setFlyout,
              closeFlyout,
              refreshDashboard,
            });
          }}
          data-test-subj={`euiLink_${alert.trigger_name}`}
        >
          {`${total} alerts`}
        </EuiLink>
      );
    },
  },
  {
    field: 'ACTIVE',
    name: 'Active',
    sortable: true,
    truncateText: false,
  },
  {
    field: 'ACKNOWLEDGED',
    name: 'Acknowledged',
    sortable: true,
    truncateText: false,
  },
  {
    field: 'ERROR',
    name: 'Errors',
    sortable: true,
    truncateText: false,
  },
  {
    field: 'trigger_name',
    name: 'Trigger name',
    sortable: true,
    truncateText: true,
    textOnly: true,
  },
  {
    field: 'start_time',
    name: 'Trigger start time',
    sortable: true,
    truncateText: false,
    render: renderTime,
    dataType: 'date',
  },
  {
    field: 'last_notification_time',
    name: 'Trigger last updated',
    sortable: true,
    truncateText: true,
    render: renderTime,
    dataType: 'date',
  },
  {
    field: 'severity',
    name: 'Severity',
    sortable: false,
    truncateText: false,
  },
  {
    field: 'monitor_name',
    name: 'Monitor name',
    sortable: true,
    truncateText: true,
    textOnly: true,
    render: (name, alert) => (
      <EuiLink href={`${PLUGIN_NAME}#/monitors/${alert.monitor_id}`}>{name}</EuiLink>
    ),
  },
];
