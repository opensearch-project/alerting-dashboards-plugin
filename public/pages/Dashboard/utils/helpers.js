/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import _ from 'lodash';
import { DEFAULT_GET_ALERTS_QUERY_PARAMS, EMPTY_ALERT_LIST, MAX_ALERT_COUNT } from './constants';
import { bucketColumns } from './tableUtils';
import { ALERT_STATE, DEFAULT_EMPTY_DATA } from '../../../utils/constants';
import queryString from 'query-string';
import { GET_ALERTS_SORT_FILTERS } from '../../../../server/services/AlertService';

export function groupAlertsByTrigger(alerts) {
  if (_.isUndefined(alerts)) return _.cloneDeep(EMPTY_ALERT_LIST.alerts);
  let alertsByTriggers = new Map();
  alerts.map((alert) => {
    const triggerID = alert.trigger_id;
    const newAlertList = alertsByTriggers.has(triggerID)
      ? addAlert(alertsByTriggers.get(triggerID), alert)
      : addFirstAlert(alert);
    alertsByTriggers.set(triggerID, newAlertList);
  });
  return Array.from(alertsByTriggers, ([triggerID, alerts]) => ({ ...alerts, triggerID }));
}

export function addFirstAlert(firstAlert) {
  const {
    state,
    version,
    trigger_name,
    severity,
    start_time,
    last_notification_time,
    monitor_name,
    monitor_id,
    workflow_id,
    workflow_name,
    alert_source,
  } = firstAlert;
  let newAlertList = _.cloneDeep(EMPTY_ALERT_LIST);
  newAlertList[state]++;
  newAlertList.total++;
  newAlertList.alerts.push(firstAlert);
  return {
    ...newAlertList,
    version,
    trigger_name,
    severity,
    start_time,
    last_notification_time,
    monitor_name: monitor_name || workflow_name,
    monitor_id: monitor_id || workflow_id,
    alert_source,
  };
}

export function addAlert(alertList, newAlert) {
  const state = newAlert.state;
  alertList[state]++;
  alertList.total++;
  alertList.alerts.push(newAlert);

  const { end_time, acknowledged_time } = newAlert;
  const alertLastUpdateTime = end_time < acknowledged_time ? acknowledged_time : end_time;
  if (alertList.last_notification_time < alertLastUpdateTime)
    alertList.last_notification_time = alertLastUpdateTime;

  return alertList;
}

export const renderEmptyValue = (value) => {
  return value === undefined ? DEFAULT_EMPTY_DATA : value;
};

export function insertGroupByColumn(groupBy = []) {
  let result = _.cloneDeep(bucketColumns);
  groupBy.map((fieldName) =>
    result.push({
      field: `agg_alert_content.bucket.key.${fieldName}`,
      name: _.capitalize(fieldName),
      render: renderEmptyValue,
      sortable: false,
      truncateText: false,
    })
  );
  return result;
}

export function removeColumns(columnFieldNames = [], allColumns = []) {
  return allColumns.filter((column) => !_.includes(columnFieldNames, column.field));
}

export function getInitialSize(
  perAlertView = false,
  defaultSize = DEFAULT_GET_ALERTS_QUERY_PARAMS.size
) {
  return perAlertView && defaultSize >= 0 ? defaultSize : MAX_ALERT_COUNT;
}

export function displayAcknowledgedAlertsToast(notifications, successfulCount = 0) {
  const successfulText = `Successfully acknowledged ${successfulCount} ${
    successfulCount === 1 ? 'alert' : 'alerts'
  }.`;
  if (successfulCount > 0) notifications.toasts.addSuccess(successfulText);
}

export function filterActiveAlerts(alerts = []) {
  return _.filter(alerts, { state: ALERT_STATE.ACTIVE });
}

export function getQueryObjectFromState({
  page,
  size,
  search,
  sortField,
  sortDirection,
  severityLevel,
  alertState,
  monitorIds,
  flyoutIsOpen,
}) {
  return {
    page,
    size,
    search,
    sortField,
    sortDirection,
    severityLevel,
    alertState,
    monitorIds,
    flyoutIsOpen,
  };
}

export function getURLQueryParams(location) {
  const {
    from = DEFAULT_GET_ALERTS_QUERY_PARAMS.from,
    size = DEFAULT_GET_ALERTS_QUERY_PARAMS.size,
    search = DEFAULT_GET_ALERTS_QUERY_PARAMS.search,
    sortField = DEFAULT_GET_ALERTS_QUERY_PARAMS.sortField,
    sortDirection = DEFAULT_GET_ALERTS_QUERY_PARAMS.sortDirection,
    severityLevel = DEFAULT_GET_ALERTS_QUERY_PARAMS.severityLevel,
    alertState = DEFAULT_GET_ALERTS_QUERY_PARAMS.alertState,
  } = queryString.parse(location.search);

  return {
    from: isNaN(parseInt(from, 10)) ? DEFAULT_GET_ALERTS_QUERY_PARAMS.from : parseInt(from, 10),
    size: isNaN(parseInt(size, 10)) ? DEFAULT_GET_ALERTS_QUERY_PARAMS.size : parseInt(size, 10),
    search,
    sortField: _.includes(_.values(GET_ALERTS_SORT_FILTERS), sortField)
      ? sortField
      : DEFAULT_GET_ALERTS_QUERY_PARAMS.sortField,
    sortDirection,
    severityLevel,
    alertState,
  };
}
