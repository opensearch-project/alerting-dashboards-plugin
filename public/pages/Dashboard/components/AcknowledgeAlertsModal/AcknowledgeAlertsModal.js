/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { Component } from 'react';
import _ from 'lodash';
import queryString from 'query-string';
import PropTypes from 'prop-types';
import {
  EuiBasicTable,
  EuiButton,
  EuiFlexGroup,
  EuiFlexItem,
  EuiHorizontalRule,
  EuiIcon,
  EuiModal,
  EuiModalBody,
  EuiModalFooter,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiOverlayMask,
} from '@elastic/eui';
import {
  ALERT_STATE,
  MONITOR_ACTIONS,
  MONITOR_GROUP_BY,
  MONITOR_INPUT_DETECTOR_ID,
  MONITOR_TYPE,
  OPENSEARCH_DASHBOARDS_AD_PLUGIN,
} from '../../../../utils/constants';
import {
  displayAcknowledgedAlertsToast,
  filterActiveAlerts,
  getQueryObjectFromState,
  getURLQueryParams,
  insertGroupByColumn,
  removeColumns,
} from '../../utils/helpers';
import { backendErrorNotification } from '../../../../utils/helpers';
import { MAX_ALERT_COUNT } from '../../utils/constants';
import { DEFAULT_PAGE_SIZE_OPTIONS } from '../../../Monitors/containers/Monitors/utils/constants';
import DashboardControls from '../DashboardControls';
import ContentPanel from '../../../../components/ContentPanel';
import { queryColumns } from '../../utils/tableUtils';
import DashboardEmptyPrompt from '../DashboardEmptyPrompt';
import { getAlertsFindingColumn } from '../FindingsDashboard/findingsUtils';

export const DEFAULT_NUM_MODAL_ROWS = 10;

export default class AcknowledgeAlertsModal extends Component {
  constructor(props) {
    super(props);
    const { location, monitor_id } = this.props;

    const {
      alertState,
      from,
      search,
      severityLevel,
      size,
      sortDirection,
      sortField,
    } = getURLQueryParams(location);

    this.state = {
      alerts: [],
      alertState: alertState,
      flyoutIsOpen: false,
      loading: true,
      monitors: [],
      monitorIds: [monitor_id],
      page: Math.floor(from / size),
      search: search,
      selectedItems: [],
      severityLevel: severityLevel,
      size: DEFAULT_NUM_MODAL_ROWS,
      sortDirection: sortDirection,
      sortField: sortField,
      totalAlerts: 0,
    };
  }

  componentDidMount() {
    const {
      alertState,
      page,
      search,
      severityLevel,
      size,
      sortDirection,
      sortField,
      monitorIds,
    } = this.state;
    this.getAlerts(
      page * size,
      size,
      search,
      sortField,
      sortDirection,
      severityLevel,
      alertState,
      monitorIds
    );
  }

  componentDidUpdate(prevProps, prevState) {
    const prevQuery = getQueryObjectFromState(prevState);
    const currQuery = getQueryObjectFromState(this.state);
    if (!_.isEqual(prevQuery, currQuery)) {
      const {
        page,
        size,
        search,
        sortField,
        sortDirection,
        severityLevel,
        alertState,
        monitorIds,
      } = this.state;
      this.getAlerts(
        page * size,
        size,
        search,
        sortField,
        sortDirection,
        severityLevel,
        alertState,
        monitorIds
      );
    }
  }

  getAlerts = async () => {
    this.setState({ ...this.state, loading: true });
    const {
      from,
      search,
      sortField,
      sortDirection,
      severityLevel,
      alertState,
      monitorIds,
    } = this.state;

    const { httpClient, history, notifications, triggerId } = this.props;

    const params = {
      from,
      size: MAX_ALERT_COUNT,
      search,
      sortField,
      sortDirection,
      severityLevel,
      alertState,
      monitorIds,
    };

    const queryParamsString = queryString.stringify(params);
    history.replace({ ...this.props.location, search: queryParamsString });

    httpClient.get('../api/alerting/alerts', { query: params }).then((resp) => {
      if (resp.ok) {
        const { alerts } = resp;
        const filteredAlerts = _.filter(alerts, { trigger_id: triggerId });
        this.setState({
          ...this.state,
          alerts: filteredAlerts,
          totalAlerts: filteredAlerts.length,
        });
      } else {
        console.log('error getting alerts:', resp);
        backendErrorNotification(notifications, 'get', 'alerts', resp.err);
      }
    });

    this.setState({ ...this.state, loading: false });
  };

  acknowledgeAlerts = async () => {
    this.setState({ loading: true });
    const { selectedItems } = this.state;
    const { httpClient, notifications } = this.props;

    if (!selectedItems.length) return;

    const selectedAlerts = filterActiveAlerts(selectedItems);

    const monitorAlerts = selectedAlerts.reduce((monitorAlerts, alert) => {
      const { id, monitor_id: monitorId } = alert;
      if (monitorAlerts[monitorId]) monitorAlerts[monitorId].push(id);
      else monitorAlerts[monitorId] = [id];
      return monitorAlerts;
    }, {});

    Object.entries(monitorAlerts).map(([monitorId, alerts]) =>
      httpClient
        .post(`../api/alerting/monitors/${monitorId}/_acknowledge/alerts`, {
          body: JSON.stringify({ alerts }),
        })
        .then((resp) => {
          if (!resp.ok) {
            backendErrorNotification(notifications, 'acknowledge', 'alert', resp.resp);
          } else {
            const successfulCount = _.get(resp, 'resp.success', []).length;
            displayAcknowledgedAlertsToast(notifications, successfulCount);
          }
        })
        .catch((error) => error)
    );

    const {
      page,
      size,
      search,
      sortField,
      sortDirection,
      severityLevel,
      alertState,
      monitorIds,
    } = this.state;
    await this.getAlerts(
      page * size,
      size,
      search,
      sortField,
      sortDirection,
      severityLevel,
      alertState,
      monitorIds
    );
    this.setState({ ...this.state, loading: false, selectedItems: [] });
  };

  onSeverityLevelChange = (e) => {
    this.setState({ page: 0, severityLevel: e.target.value });
  };

  onAlertStateChange = (e) => {
    this.setState({ page: 0, alertState: e.target.value });
  };

  onPageClick = (page) => {
    this.setState({ page });
  };

  onSearchChange = (e) => {
    this.setState({ page: 0, search: e.target.value });
  };

  onSelectionChange = (selectedItems) => {
    this.setState({ selectedItems });
  };

  onTableChange = ({ page: tablePage = {}, sort = {} }) => {
    const { index: page, size } = tablePage;

    const { field: sortField, direction: sortDirection } = sort;
    this.setState({
      page,
      size,
      sortField,
      sortDirection,
    });

    const { alerts } = this.props;
    this.setState({ alerts });
  };

  onCreateTrigger = () => {
    const { history, monitorId, onClose } = this.props;
    onClose();
    history.push(`/monitors/${monitorId}?action=${MONITOR_ACTIONS.UPDATE_MONITOR}`);
  };

  render() {
    const {
      httpClient,
      location,
      history,
      monitor,
      notifications,
      onClose,
      triggerName,
    } = this.props;
    const detectorId = _.get(monitor, MONITOR_INPUT_DETECTOR_ID);
    const groupBy = _.get(monitor, MONITOR_GROUP_BY);
    const monitorType = _.get(monitor, 'monitor_type', MONITOR_TYPE.QUERY_LEVEL);

    const actions = () => {
      const { selectedItems } = this.state;
      const actions = [
        <EuiButton
          onClick={this.acknowledgeAlerts}
          disabled={!selectedItems.length}
          data-test-subj={'alertsDashboardModal_acknowledgeAlertsButton'}
        >
          Acknowledge
        </EuiButton>,
      ];
      if (!_.isEmpty(detectorId)) {
        actions.unshift(
          <EuiButton
            href={`${OPENSEARCH_DASHBOARDS_AD_PLUGIN}#/detectors/${detectorId}`}
            target="_blank"
          >
            View detector <EuiIcon type="popout" />
          </EuiButton>
        );
      }
      return actions;
    };

    const getItemId = (item) => {
      switch (monitorType) {
        case MONITOR_TYPE.QUERY_LEVEL:
        case MONITOR_TYPE.CLUSTER_METRICS:
        case MONITOR_TYPE.DOC_LEVEL:
          return `${item.id}-${item.version}`;
        case MONITOR_TYPE.BUCKET_LEVEL:
          return item.id;
      }
    };

    const {
      alerts = [],
      alertState,
      flyoutIsOpen,
      loading,
      page,
      search,
      selectable,
      selectedItems,
      severityLevel,
      size,
      sortDirection,
      sortField,
      totalAlerts,
    } = this.state;

    const columnType = () => {
      let columns;
      switch (monitorType) {
        case MONITOR_TYPE.BUCKET_LEVEL:
          columns = insertGroupByColumn(groupBy);
          break;
        case MONITOR_TYPE.DOC_LEVEL:
          columns = _.cloneDeep(queryColumns);
          columns.splice(
            0,
            0,
            getAlertsFindingColumn(
              httpClient,
              history,
              location,
              notifications,
              flyoutIsOpen,
              () => this.setState({ flyoutIsOpen: true }),
              () => this.setState({ flyoutIsOpen: false })
            )
          );
          break;
        default:
          columns = queryColumns;
          break;
      }
      return removeColumns(['trigger_name'], columns);
    };

    const pagination = {
      pageIndex: page,
      pageSize: size,
      totalItemCount: totalAlerts,
      pageSizeOptions: DEFAULT_PAGE_SIZE_OPTIONS,
    };

    const selection = {
      onSelectionChange: this.onSelectionChange,
      selectable: (item) => item.state === ALERT_STATE.ACTIVE,
      selectableMessage: (selectable) =>
        selectable ? undefined : 'Only active alerts can be acknowledged.',
    };

    const sorting = {
      sort: {
        direction: sortDirection,
        field: sortField,
      },
    };

    const trimmedAlerts = alerts.slice(page * size, page * size + size);

    return (
      <EuiOverlayMask>
        <EuiModal
          onClose={onClose}
          maxWidth={false}
          data-test-subj={`alertsDashboardModal_${triggerName}`}
        >
          <EuiModalHeader data-test-subj={`alertsDashboardModal_header_${triggerName}`}>
            <EuiModalHeaderTitle>
              {`Select which alerts to acknowledge for ${triggerName}`}
            </EuiModalHeaderTitle>
          </EuiModalHeader>
          <EuiModalBody>
            <EuiFlexGroup>
              <EuiFlexItem>
                <ContentPanel
                  title={`Alerts (${selectedItems.length}/${totalAlerts})`}
                  titleSize={'s'}
                  bodyStyles={{ padding: 'initial' }}
                  actions={actions()}
                >
                  <DashboardControls
                    activePage={page}
                    pageCount={Math.ceil(totalAlerts / size) || 1}
                    search={search}
                    severity={severityLevel}
                    state={alertState}
                    onSearchChange={this.onSearchChange}
                    onSeverityChange={this.onSeverityLevelChange}
                    onStateChange={this.onAlertStateChange}
                    onPageChange={this.onPageClick}
                  />

                  <EuiHorizontalRule margin="xs" />

                  <EuiBasicTable
                    items={loading ? [] : trimmedAlerts}
                    /*
                     * If using just ID, doesn't update selectedItems when doing acknowledge
                     * because the next getAlerts have the same id
                     * $id-$version will correctly remove selected items
                     * */
                    itemId={getItemId}
                    columns={columnType()}
                    loading={loading}
                    pagination={pagination}
                    sorting={sorting}
                    isSelectable={selectable}
                    selection={selection}
                    onChange={this.onTableChange}
                    noItemsMessage={
                      loading ? (
                        'Loading alerts...'
                      ) : (
                        <DashboardEmptyPrompt
                          onCreateTrigger={this.onCreateTrigger}
                          isModal={true}
                        />
                      )
                    }
                    data-test-subj={`alertsDashboardModal_table_${triggerName}`}
                  />
                </ContentPanel>
              </EuiFlexItem>
            </EuiFlexGroup>
          </EuiModalBody>
          <EuiModalFooter>
            <EuiButton
              onClick={onClose}
              fill
              data-test-subj={`alertsDashboardModal_closeButton_${triggerName}`}
            >
              Close
            </EuiButton>
          </EuiModalFooter>
        </EuiModal>
      </EuiOverlayMask>
    );
  }
}

AcknowledgeAlertsModal.propTypes = {
  httpClient: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  location: PropTypes.object.isRequired,
  monitor: PropTypes.object.isRequired,
  monitorId: PropTypes.string.isRequired,
  notifications: PropTypes.object.isRequired,
  triggerId: PropTypes.string.isRequired,
  triggerName: PropTypes.string.isRequired,
  onClose: PropTypes.func.isRequired,
};
