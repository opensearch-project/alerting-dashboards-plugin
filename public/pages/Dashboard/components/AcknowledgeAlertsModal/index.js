/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import AcknowledgeAlertsModal from './AcknowledgeAlertsModal';

export default AcknowledgeAlertsModal;
