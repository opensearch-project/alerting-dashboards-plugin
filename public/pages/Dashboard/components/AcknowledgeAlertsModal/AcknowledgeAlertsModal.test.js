/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import _ from 'lodash';
import { historyMock, httpClientMock } from '../../../../../test/mocks';
import AcknowledgeAlertsModal from './AcknowledgeAlertsModal';
import { FORMIK_INITIAL_VALUES } from '../../../CreateMonitor/containers/CreateMonitor/utils/constants';
import coreMock from '../../../../../test/mocks/CoreMock';

beforeEach(() => {
  jest.clearAllMocks();
});

describe('AcknowledgeAlertsModal', () => {
  test('renders', () => {
    const location = {
      hash: '',
      pathname: '/dashboard',
      search: '',
      state: undefined,
    };
    const notifications = _.cloneDeep(coreMock.notifications);
    const component = (
      <AcknowledgeAlertsModal
        httpClient={httpClientMock}
        history={historyMock}
        location={location}
        monitor={FORMIK_INITIAL_VALUES}
        monitorId={'fake_monitor_id'}
        notifications={notifications}
        triggerId={'fake_trigger_id'}
        triggerName={'fake_trigger_name'}
        onClose={() => {}}
      />
    );
    expect(component).toMatchSnapshot();
  });
});
