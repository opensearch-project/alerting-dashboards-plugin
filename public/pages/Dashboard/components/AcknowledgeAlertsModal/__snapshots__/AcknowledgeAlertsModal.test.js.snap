// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AcknowledgeAlertsModal renders 1`] = `
<AcknowledgeAlertsModal
  history={
    Object {
      "action": "REPLACE",
      "block": [MockFunction],
      "createHref": [MockFunction],
      "go": [MockFunction],
      "goBack": [MockFunction],
      "goForward": [MockFunction],
      "length": 0,
      "listen": [MockFunction],
      "location": Object {
        "hash": "",
        "pathname": "",
        "search": "",
        "state": undefined,
      },
      "push": [MockFunction],
      "replace": [MockFunction],
    }
  }
  httpClient={[MockFunction]}
  location={
    Object {
      "hash": "",
      "pathname": "/dashboard",
      "search": "",
      "state": undefined,
    }
  }
  monitor={
    Object {
      "aggregationType": "count",
      "aggregations": Array [],
      "associatedMonitors": Object {
        "sequence": Object {
          "delegates": Array [],
        },
      },
      "associatedMonitorsEditor": "",
      "associatedMonitorsList": Array [],
      "bucketUnitOfTime": "h",
      "bucketValue": 1,
      "clusterNames": Array [],
      "cronExpression": "0 */1 * * *",
      "daily": 0,
      "description": "",
      "detectorId": "",
      "disabled": false,
      "fieldName": Array [],
      "filters": Array [],
      "frequency": "interval",
      "groupBy": Array [],
      "groupByField": Array [
        Object {
          "label": "",
        },
      ],
      "groupedOverFieldName": "bytes",
      "groupedOverTop": 5,
      "index": Array [],
      "monitor_type": "query_level_monitor",
      "monthly": Object {
        "day": 1,
        "type": "day",
      },
      "name": "",
      "overDocuments": "all documents",
      "period": Object {
        "interval": 1,
        "unit": "MINUTES",
      },
      "preventVisualEditor": false,
      "queries": Array [],
      "query": "{
    \\"size\\": 0,
    \\"query\\": {
        \\"match_all\\": {}
    }
}",
      "searchType": "graph",
      "timeField": "",
      "timezone": Array [],
      "uri": Object {
        "api_type": "",
        "clusters": Array [],
        "path": "",
        "path_params": "",
        "url": "",
      },
      "weekly": Object {
        "fri": false,
        "mon": false,
        "sat": false,
        "sun": false,
        "thur": false,
        "tue": false,
        "wed": false,
      },
    }
  }
  monitorId="fake_monitor_id"
  notifications={
    Object {
      "toasts": Object {
        "addDanger": [MockFunction addDanger],
        "addSuccess": [MockFunction addSuccess],
      },
    }
  }
  onClose={[Function]}
  triggerId="fake_trigger_id"
  triggerName="fake_trigger_name"
/>
`;
